parsing/src/main/scala/org/http4s/parsing/Rfc7230.scala

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2021 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.parsing
+
+import cats.parse.Parser
+import cats.parse.Parser.{char, string}
+import cats.parse.Rfc5234.digit
+
+object Rfc7230 {
+  /* @see [[https://tools.ietf.org/html/rfc7230#section-2.6 RFC7230, Section 2.6]]*/
+  val httpName =
+    // HTTP-name = %x48.54.54.50 ; "HTTP", case-sensitive
+    string("HTTP")
+
+  /* @see [[https://tools.ietf.org/html/rfc7230#section-2.6 RFC7230, Section 2.6]]*/
+  val httpVersion: Parser[(Int, Int)] =
```

Comment:
This is where the lack of http4s-core hurts.

We can use anonymous tuples like this, which isn't really so bad for two, but is going to hit hard on more complex types.  Or we could design a set of anemic case classes, but that will create more garbage: from tuple to case class, and from case class to consuming library's richer model.

Reply:
I don't have good idea of what http4s core would do, so this is probably a silly question: Could Scala3's opaque types help?

Reply:
Eh, maybe?  We'd have an extra map call.  I don't know if we can cast from a representation to an opaque type and avoid an allocation, so it might effectively be just a `.map(identity)`.

Bigger problem is that would be a Scala 3 solution, and at least http4s will support Scala 2 for some time to come.

Reply:
Re, Scala 2, I read on the [Scala 3 Source level page](https://scalacenter.github.io/scala-3-migration-guide/docs/compatibility/source.html) that
>  But amazingly, this new Scala 3 artifact can be consumed as a dependency in Scala 2.13.

Reply:
@rossabaker what if we introduced something like:
```scala
trait Rfc7230[M] {
  def create(major, minor): M
}
```

And then define something like:
```scala
def httpVersionParser[M: Rfc7230]: Parser[M] = 
  httpVersion.map { case (major, minor) => Rfc7230[M].create(major, minor) }
```

Along with this in core:
```scala
implicit object rfc7230ForHttpVersion extends Rfc7230[HttpVersion] {
  def create(major: Int, minor: Int) = new HttpVersion(major, minor)
}
```

Reply:
And btw for extra points I think this optimization even avoids the extra lambda?

```scala
trait Rfc7230[M] extends ((Int, Int) => M) {
  final def apply(x: (Int, Int)): M = parse(x._1, x._2)
  def create(major, minor): M
}

def httpVersionParser[M](implicit toModel: Rfc7230[M]): Parser[M] = 
  httpVersion.map(toModel)
```